receiver/k8sobjectsreceiver/receiver_references_test.go

```diff
@@ -0,0 +1,176 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8sobjectsreceiver
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"testing"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"go.opentelemetry.io/collector/component/componenttest"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/receiver/receivertest"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+func TestReferences(t *testing.T) {
+	RegisterFailHandler(Fail)
+
+	RunSpecs(t, "Receiver References tests")
+}
+
+var _ = Context("Lumigo controller", func() {
+
+	It("Send all needed Pod references", func() {
+		mockClient := newMockDynamicClient()
+
+		rCfg := createDefaultConfig().(*Config)
+		rCfg.makeDynamicClient = mockClient.getMockDynamicClient
+		rCfg.makeDiscoveryClient = getMockDiscoveryClient
+
+		rCfg.Objects = []*K8sObjectsConfig{
+			{
+				Name:       "pods",
+				Mode:       WatchMode,
+				Namespaces: []string{"default"},
+			},
+			{
+				Name:       "events",
+				Mode:       WatchMode,
+				Namespaces: []string{"default"},
+			},
+		}
+
+		err := rCfg.Validate()
+		Expect(err).NotTo(HaveOccurred())
+
+		consumer := newMockLogConsumer()
+		r, err := newReceiver(
+			receivertest.NewNopCreateSettings(),
+			rCfg,
+			consumer,
+		)
+		Expect(r).NotTo(BeNil())
+		Expect(err).NotTo(HaveOccurred())
+
+		ctx := context.Background()
+		Expect(r.Start(ctx, componenttest.NewNopHost())).NotTo(HaveOccurred())
+
+		time.Sleep(time.Millisecond * 100)
+
+		pod1 := generatePod("pod1", "default", map[string]interface{}{
+			"environment": "production",
+		}, "1234")
+		pod2 := generatePod("pod2", "default", map[string]interface{}{
+			"environment": "test",
+		}, "5678")
+		pod3 := generatePod("pod3", "default_ignore", map[string]interface{}{
+			"environment": "production",
+		}, "9012")
+		mockClient.createPods(pod1, pod2, pod3)
+		time.Sleep(time.Second * 1)
+
+		// The pods mentioned by the events below have all been reported yet.
+		// We expect only one log per event.
+		mockClient.createEvents(
+			generateEvent("event1", "default", map[string]interface{}{
+				"environment": "production",
+			}, pod1),
+			generateEvent("event2", "default", map[string]interface{}{
+				"environment": "production",
+			}, pod2),
+		)
+
+		time.Sleep(time.Second * 1)
+
+		Eventually(func(g Gomega) {
+			logRecords := getLogRecords(consumer.Logs())
+
+			g.Expect(logRecords).Should(ContainElement(matchWatchEvent(watch.Added, toObjectReference(pod1))))
+			g.Expect(logRecords).Should(ContainElement(matchWatchEvent(watch.Added, toObjectReference(pod2))))
+		}, 1*time.Minute, 1*time.Second).Should(Succeed())
+
+		// TODO Improve tests using a deployment and check that all needed resourceVersions of pods and replicasets
```

Comment:
😳

Reply:
I have not focused on the unit tests in the collector, I focused on the end-to-end tests of the `lumigo-kubernetes-operator` :D